File: engines/unreal/android/src/main/kotlin/com/xraph/gameframework/unreal/UnrealEngineController.kt
package com.xraph.gameframework.unreal

import android.app.Activity
import android.content.Context
import android.view.View
import android.view.ViewGroup
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Unreal Engine Controller for Android
 *
 * Manages the Unreal Engine lifecycle, view integration, and communication
 * between Flutter and Unreal Engine on Android.
 */
class UnrealEngineController(
    private val context: Context,
    private val activity: Activity,
    private val viewId: Int,
    private val channel: MethodChannel,
    private val config: Map<String, Any>
) {
    private var unrealView: View? = null
    private val isReady = AtomicBoolean(false)
    private val isPaused = AtomicBoolean(false)
    private val isDestroyed = AtomicBoolean(false)

    companion object {
        private const val TAG = "UnrealEngineController"
        const val ENGINE_TYPE = "unreal"
        const val ENGINE_VERSION = "5.3.0"

        init {
            try {
                System.loadLibrary("UnrealFlutterBridge")
            } catch (e: UnsatisfiedLinkError) {
                // Library not available - Unreal native library not linked
                android.util.Log.w(TAG, "UnrealFlutterBridge native library not found")
            }
        }
    }

    // MARK: - Lifecycle Methods

    /**
     * Initialize and create the Unreal Engine instance
     */
    fun create(): Boolean {
        if (isDestroyed.get()) {
            sendError("Cannot create destroyed engine")
            return false
        }

        if (isReady.get()) {
            return true
        }

        try {
            // Initialize Unreal Engine native
            if (!nativeCreate(config)) {
                sendError("Failed to create Unreal Engine instance")
                return false
            }

            // Create Unreal view
            unrealView = nativeGetView()
            if (unrealView == null) {
                sendError("Failed to get Unreal view")
                return false
            }

            isReady.set(true)
            sendEvent("created")
            sendEvent("loaded")

            return true
        } catch (e: Exception) {
            sendError("Exception during engine creation: ${e.message}")
            return false
        }
    }

    /**
     * Pause the Unreal Engine
     */
    fun pause() {
        if (!isReady.get() || isDestroyed.get()) {
            return
        }

        try {
            nativePause()
            isPaused.set(true)
            sendEvent("paused")
        } catch (e: Exception) {
            sendError("Exception during pause: ${e.message}")
        }
    }

    /**
     * Resume the Unreal Engine
     */
    fun resume() {
        if (!isReady.get() || isDestroyed.get()) {
            return
        }

        try {
            nativeResume()
            isPaused.set(false)
            sendEvent("resumed")
        } catch (e: Exception) {
            sendError("Exception during resume: ${e.message}")
        }
    }

    /**
     * Unload the Unreal Engine (pause and detach)
     */
    fun unload() {
        if (!isReady.get() || isDestroyed.get()) {
            return
        }

        try {
            pause()
            sendEvent("unloaded")
        } catch (e: Exception) {
            sendError("Exception during unload: ${e.message}")
        }
    }

    /**
     * Quit and destroy the Unreal Engine
     */
    fun quit() {
        if (isDestroyed.get()) {
            return
        }

        try {
            nativeQuit()
            unrealView = null
            isReady.set(false)
            isDestroyed.set(true)
            sendEvent("destroyed")
        } catch (e: Exception) {
            sendError("Exception during quit: ${e.message}")
        }
    }

    // MARK: - Communication Methods

    /**
     * Send a message to Unreal Engine
     */
    fun sendMessage(target: String, method: String, data: String) {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready for messages")
            return
        }

        try {
            nativeSendMessage(target, method, data)
        } catch (e: Exception) {
            sendError("Failed to send message: ${e.message}")
        }
    }

    /**
     * Send a JSON message to Unreal Engine
     */
    fun sendJsonMessage(target: String, method: String, data: Map<String, Any>) {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready for messages")
            return
        }

        try {
            val jsonString = org.json.JSONObject(data).toString()
            nativeSendMessage(target, method, jsonString)
        } catch (e: Exception) {
            sendError("Failed to send JSON message: ${e.message}")
        }
    }

    // MARK: - Unreal-Specific Methods

    /**
     * Execute a console command in Unreal Engine
     */
    fun executeConsoleCommand(command: String) {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready for console commands")
            return
        }

        try {
            nativeExecuteConsoleCommand(command)
        } catch (e: Exception) {
            sendError("Failed to execute console command: ${e.message}")
        }
    }

    /**
     * Load a level/map in Unreal Engine
     */
    fun loadLevel(levelName: String) {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready to load level")
            return
        }

        try {
            nativeLoadLevel(levelName)
        } catch (e: Exception) {
            sendError("Failed to load level: ${e.message}")
        }
    }

    /**
     * Apply quality settings to Unreal Engine
     */
    fun applyQualitySettings(settings: Map<String, Any>) {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready for quality settings")
            return
        }

        try {
            nativeApplyQualitySettings(settings)
        } catch (e: Exception) {
            sendError("Failed to apply quality settings: ${e.message}")
        }
    }

    /**
     * Get current quality settings from Unreal Engine
     */
    fun getQualitySettings(): Map<String, Any>? {
        if (!isReady.get() || isDestroyed.get()) {
            sendError("Engine not ready")
            return null
        }

        return try {
            nativeGetQualitySettings()
        } catch (e: Exception) {
            sendError("Failed to get quality settings: ${e.message}")
            null
        }
    }

    /**
     * Check if engine is in background
     */
    fun isInBackground(): Boolean {
        return isPaused.get()
    }

    // MARK: - View Integration

    /**
     * Get the Unreal Engine view to attach to Flutter
     */
    fun getView(): View? {
        return unrealView
    }

    /**
     * Attach Unreal view to parent
     */
    fun attachView(parent: ViewGroup) {
        unrealView?.let { view ->
            if (view.parent == null) {
                parent.addView(view)
                sendEvent("attached")
            }
        }
    }

    /**
     * Detach Unreal view from parent
     */
    fun detachView() {
        unrealView?.let { view ->
            (view.parent as? ViewGroup)?.removeView(view)
            sendEvent("detached")
        }
    }

    // MARK: - Event Handling

    private fun sendEvent(eventType: String, message: String? = null) {
        activity.runOnUiThread {
            channel.invokeMethod("onEvent", mapOf(
                "type" to eventType,
                "message" to message
            ))
        }
    }

    private fun sendError(message: String) {
        android.util.Log.e(TAG, message)
        sendEvent("error", message)
    }

    /**
     * Called from native code when a message is received from Unreal
     */
    @Suppress("unused")
    fun onMessageFromUnreal(target: String, method: String, data: String) {
        activity.runOnUiThread {
            channel.invokeMethod("onMessage", mapOf(
                "target" to target,
                "method" to method,
                "data" to data
            ))
        }
    }

    /**
     * Called from native code when a level is loaded
     */
    @Suppress("unused")
    fun onLevelLoaded(levelName: String, buildIndex: Int) {
        activity.runOnUiThread {
            channel.invokeMethod("onLevelLoaded", mapOf(
                "name" to levelName,
                "buildIndex" to buildIndex,
                "isLoaded" to true,
                "isValid" to true,
                "metadata" to emptyMap<String, Any>()
            ))
        }
    }

    // MARK: - Native Methods (JNI)

    /**
     * Create Unreal Engine instance
     */
    private external fun nativeCreate(config: Map<String, Any>): Boolean

    /**
     * Get the native Unreal view
     */
    private external fun nativeGetView(): View?

    /**
     * Pause the engine
     */
    private external fun nativePause()

    /**
     * Resume the engine
     */
    private external fun nativeResume()

    /**
     * Quit the engine
     */
    private external fun nativeQuit()

    /**
     * Send message to Unreal
     */
    private external fun nativeSendMessage(target: String, method: String, data: String)

    /**
     * Execute console command
     */
    private external fun nativeExecuteConsoleCommand(command: String)

    /**
     * Load level
     */
    private external fun nativeLoadLevel(levelName: String)

    /**
     * Apply quality settings
     */
    private external fun nativeApplyQualitySettings(settings: Map<String, Any>)

    /**
     * Get quality settings
     */
    private external fun nativeGetQualitySettings(): Map<String, Any>
}

File: engines/unreal/ios/Classes/UnrealEngineController.swift
import UIKit
import Flutter

/**
 * Unreal Engine Controller for iOS
 *
 * Manages the Unreal Engine lifecycle, view integration, and communication
 * between Flutter and Unreal Engine on iOS.
 */
public class UnrealEngineController: NSObject {

    // MARK: - Properties

    private let viewId: Int
    private let channel: FlutterMethodChannel
    private let config: [String: Any]

    private var unrealView: UIView?
    private var unrealFramework: UnrealFramework?

    private var isReady: Bool = false
    private var isPaused: Bool = false
    private var isDestroyed: Bool = false

    // MARK: - Constants

    private static let engineType = "unreal"
    private static let engineVersion = "5.3.0"

    // MARK: - Initialization

    public init(viewId: Int, channel: FlutterMethodChannel, config: [String: Any]) {
        self.viewId = viewId
        self.channel = channel
        self.config = config
        super.init()
    }

    // MARK: - Lifecycle Methods

    /**
     * Initialize and create the Unreal Engine instance
     */
    public func create() -> Bool {
        if isDestroyed {
            sendError("Cannot create destroyed engine")
            return false
        }

        if isReady {
            return true
        }

        // Load Unreal Framework
        guard let framework = loadUnrealFramework() else {
            sendError("Failed to load Unreal Framework")
            return false
        }

        unrealFramework = framework

        // Apply configuration
        applyConfiguration(config)

        // Initialize Unreal Engine
        if !nativeCreate(config) {
            sendError("Failed to create Unreal Engine instance")
            return false
        }

        // Get Unreal view
        guard let view = nativeGetView() else {
            sendError("Failed to get Unreal view")
            return false
        }

        unrealView = view
        isReady = true

        sendEvent("created")
        sendEvent("loaded")

        return true
    }

    /**
     * Pause the Unreal Engine
     */
    public func pause() {
        if !isReady || isDestroyed {
            return
        }

        nativePause()
        isPaused = true
        sendEvent("paused")
    }

    /**
     * Resume the Unreal Engine
     */
    public func resume() {
        if !isReady || isDestroyed {
            return
        }

        nativeResume()
        isPaused = false
        sendEvent("resumed")
    }

    /**
     * Unload the Unreal Engine (pause and detach)
     */
    public func unload() {
        if !isReady || isDestroyed {
            return
        }

        pause()
        sendEvent("unloaded")
    }

    /**
     * Quit and destroy the Unreal Engine
     */
    public func quit() {
        if isDestroyed {
            return
        }

        nativeQuit()
        unrealView = nil
        unrealFramework = nil
        isReady = false
        isDestroyed = true
        sendEvent("destroyed")
    }

    // MARK: - Communication Methods

    /**
     * Send a message to Unreal Engine
     */
    public func sendMessage(target: String, method: String, data: String) {
        if !isReady || isDestroyed {
            sendError("Engine not ready for messages")
            return
        }

        nativeSendMessage(target: target, method: method, data: data)
    }

    /**
     * Send a JSON message to Unreal Engine
     */
    public func sendJsonMessage(target: String, method: String, data: [String: Any]) {
        if !isReady || isDestroyed {
            sendError("Engine not ready for messages")
            return
        }

        guard let jsonData = try? JSONSerialization.data(withJSONObject: data, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            sendError("Failed to serialize JSON message")
            return
        }

        nativeSendMessage(target: target, method: method, data: jsonString)
    }

    // MARK: - Unreal-Specific Methods

    /**
     * Execute a console command in Unreal Engine
     */
    public func executeConsoleCommand(_ command: String) {
        if !isReady || isDestroyed {
            sendError("Engine not ready for console commands")
            return
        }

        nativeExecuteConsoleCommand(command)
    }

    /**
     * Load a level/map in Unreal Engine
     */
    public func loadLevel(_ levelName: String) {
        if !isReady || isDestroyed {
            sendError("Engine not ready to load level")
            return
        }

        nativeLoadLevel(levelName)
    }

    /**
     * Apply quality settings to Unreal Engine
     */
    public func applyQualitySettings(_ settings: [String: Any]) {
        if !isReady || isDestroyed {
            sendError("Engine not ready for quality settings")
            return
        }

        nativeApplyQualitySettings(settings)
    }

    /**
     * Get current quality settings from Unreal Engine
     */
    public func getQualitySettings() -> [String: Any]? {
        if !isReady || isDestroyed {
            sendError("Engine not ready")
            return nil
        }

        return nativeGetQualitySettings()
    }

    /**
     * Check if engine is in background
     */
    public func isInBackground() -> Bool {
        return isPaused
    }

    // MARK: - View Integration

    /**
     * Get the Unreal Engine view to attach to Flutter
     */
    public func getView() -> UIView? {
        return unrealView
    }

    /**
     * Attach Unreal view to parent
     */
    public func attachView(to parent: UIView) {
        guard let view = unrealView, view.superview == nil else {
            return
        }

        view.frame = parent.bounds
        view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        parent.addSubview(view)
        sendEvent("attached")
    }

    /**
     * Detach Unreal view from parent
     */
    public func detachView() {
        guard let view = unrealView else {
            return
        }

        view.removeFromSuperview()
        sendEvent("detached")
    }

    // MARK: - Event Handling

    private func sendEvent(_ eventType: String, message: String? = nil) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            var arguments: [String: Any] = ["type": eventType]
            if let message = message {
                arguments["message"] = message
            }

            self.channel.invokeMethod("onEvent", arguments: arguments)
        }
    }

    private func sendError(_ message: String) {
        NSLog("[UnrealEngineController] Error: \(message)")
        sendEvent("error", message: message)
    }

    /**
     * Called from native code when a message is received from Unreal
     */
    @objc public func onMessageFromUnreal(target: String, method: String, data: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.channel.invokeMethod("onMessage", arguments: [
                "target": target,
                "method": method,
                "data": data
            ])
        }
    }

    /**
     * Called from native code when a level is loaded
     */
    @objc public func onLevelLoaded(levelName: String, buildIndex: Int) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.channel.invokeMethod("onLevelLoaded", arguments: [
                "name": levelName,
                "buildIndex": buildIndex,
                "isLoaded": true,
                "isValid": true,
                "metadata": [String: Any]()
            ])
        }
    }

    // MARK: - Framework Loading

    private func loadUnrealFramework() -> UnrealFramework? {
        // Load Unreal Framework from app bundle
        let bundlePath = Bundle.main.bundlePath + "/Frameworks/UnrealFramework.framework"

        guard let bundle = Bundle(path: bundlePath) else {
            NSLog("[UnrealEngineController] Failed to find UnrealFramework bundle at: \(bundlePath)")
            return nil
        }

        guard bundle.load() else {
            NSLog("[UnrealEngineController] Failed to load UnrealFramework bundle")
            return nil
        }

        NSLog("[UnrealEngineController] Successfully loaded UnrealFramework")
        return UnrealFramework(bundle: bundle)
    }

    private func applyConfiguration(_ config: [String: Any]) {
        // Apply configuration to Unreal Engine
        // This can include graphics settings, game mode, etc.
        if let enableMetal = config["enableMetal"] as? Bool, enableMetal {
            // Enable Metal graphics API
            NSLog("[UnrealEngineController] Metal graphics enabled")
        }

        if let enableHighDPI = config["enableHighDPI"] as? Bool, enableHighDPI {
            // Enable high DPI rendering
            NSLog("[UnrealEngineController] High DPI rendering enabled")
        }
    }

    // MARK: - Native Bridge Methods (Objective-C++)

    /**
     * Create Unreal Engine instance
     * Implemented in Objective-C++ bridge
     */
    private func nativeCreate(_ config: [String: Any]) -> Bool {
        // This will call into Objective-C++ bridge -> Unreal C++
        return UnrealBridge.shared.create(config: config, controller: self)
    }

    /**
     * Get the native Unreal view
     * Implemented in Objective-C++ bridge
     */
    private func nativeGetView() -> UIView? {
        return UnrealBridge.shared.getView()
    }

    /**
     * Pause the engine
     * Implemented in Objective-C++ bridge
     */
    private func nativePause() {
        UnrealBridge.shared.pause()
    }

    /**
     * Resume the engine
     * Implemented in Objective-C++ bridge
     */
    private func nativeResume() {
        UnrealBridge.shared.resume()
    }

    /**
     * Quit the engine
     * Implemented in Objective-C++ bridge
     */
    private func nativeQuit() {
        UnrealBridge.shared.quit()
    }

    /**
     * Send message to Unreal
     * Implemented in Objective-C++ bridge
     */
    private func nativeSendMessage(target: String, method: String, data: String) {
        UnrealBridge.shared.sendMessage(target: target, method: method, data: data)
    }

    /**
     * Execute console command
     * Implemented in Objective-C++ bridge
     */
    private func nativeExecuteConsoleCommand(_ command: String) {
        UnrealBridge.shared.executeConsoleCommand(command)
    }

    /**
     * Load level
     * Implemented in Objective-C++ bridge
     */
    private func nativeLoadLevel(_ levelName: String) {
        UnrealBridge.shared.loadLevel(levelName)
    }

    /**
     * Apply quality settings
     * Implemented in Objective-C++ bridge
     */
    private func nativeApplyQualitySettings(_ settings: [String: Any]) {
        UnrealBridge.shared.applyQualitySettings(settings)
    }

    /**
     * Get quality settings
     * Implemented in Objective-C++ bridge
     */
    private func nativeGetQualitySettings() -> [String: Any] {
        return UnrealBridge.shared.getQualitySettings()
    }
}

// MARK: - Unreal Framework Wrapper

/**
 * Wrapper for Unreal Framework bundle
 */
private class UnrealFramework {
    let bundle: Bundle

    init(bundle: Bundle) {
        self.bundle = bundle
    }
}

// MARK: - Unreal Bridge Interface

/**
 * Bridge to Objective-C++ code that interfaces with Unreal C++
 * This class will be implemented in UnrealBridge.mm (Objective-C++)
 */
@objc public class UnrealBridge: NSObject {

    @objc public static let shared = UnrealBridge()

    private weak var controller: UnrealEngineController?

    private override init() {
        super.init()
    }

    // These methods will be implemented in UnrealBridge.mm (Objective-C++)
    // They serve as the interface between Swift and Unreal C++

    @objc public func create(config: [String: Any], controller: UnrealEngineController) -> Bool {
        self.controller = controller
        // Implementation in UnrealBridge.mm
        return false // Placeholder
    }

    @objc public func getView() -> UIView? {
        // Implementation in UnrealBridge.mm
        return nil // Placeholder
    }

    @objc public func pause() {
        // Implementation in UnrealBridge.mm
    }

    @objc public func resume() {
        // Implementation in UnrealBridge.mm
    }

    @objc public func quit() {
        // Implementation in UnrealBridge.mm
    }

    @objc public func sendMessage(target: String, method: String, data: String) {
        // Implementation in UnrealBridge.mm
    }

    @objc public func executeConsoleCommand(_ command: String) {
        // Implementation in UnrealBridge.mm
    }

    @objc public func loadLevel(_ levelName: String) {
        // Implementation in UnrealBridge.mm
    }

    @objc public func applyQualitySettings(_ settings: [String: Any]) {
        // Implementation in UnrealBridge.mm
    }

    @objc public func getQualitySettings() -> [String: Any] {
        // Implementation in UnrealBridge.mm
        return [:] // Placeholder
    }

    // Called from Unreal C++ to send messages to Flutter
    @objc public func notifyMessage(target: String, method: String, data: String) {
        controller?.onMessageFromUnreal(target: target, method: method, data: data)
    }

    // Called from Unreal C++ when level is loaded
    @objc public func notifyLevelLoaded(levelName: String, buildIndex: Int) {
        controller?.onLevelLoaded(levelName: levelName, buildIndex: buildIndex)
    }
}
